pkg/controller/node/cidr_allocator.go

```diff
@@ -0,0 +1,182 @@
+type emptyAllocator struct{}
+
+// AllocateNext always returns nil and error
+func (r *emptyAllocator) AllocateNext() (cidr *net.IPNet, err error) {
+	return nil, errCIDRRangeNoCIDRsRemaining
+}
+
+// Release is a noop
+func (r *emptyAllocator) Release(_ *net.IPNet) {
+	// noop
+}
+
+type rangeAllocator struct {
+	clusterCIDR    *net.IPNet
+	subNetMaskSize int
+	cidrs          chan *net.IPNet
+	used           big.Int
+	lock           sync.Mutex
+	rand           *rand.Rand
+}
+
+// NewCIDRRangeAllocator returns a CIDRAllocator to allocate CIDR for node
+// If subNetMaskSize is less than cluster CIDR mask size, an emptyAllocator is returned.
+// Otherwise an rangeAllocator is returned.
+func NewCIDRRangeAllocator(clusterCIDR *net.IPNet, subNetMaskSize int) CIDRAllocator {
+	cidrMask := clusterCIDR.Mask
+	cidrMaskSize, _ := cidrMask.Size()
+
+	if subNetMaskSize < cidrMaskSize {
+		return &emptyAllocator{}
+	}
+
+	ra := &rangeAllocator{
+		clusterCIDR:    clusterCIDR,
+		subNetMaskSize: subNetMaskSize,
+		cidrs:          make(chan *net.IPNet, cidrsBufSize),
+		rand:           rand.New(rand.NewSource(time.Now().UnixNano())),
+	}
+	go util.Until(func() { ra.fillCIDRs(util.NeverStop) }, nextFreeCIDRCooldown, util.NeverStop)
+	return ra
+}
+
+func (r *rangeAllocator) AllocateNext() (cidr *net.IPNet, err error) {
+	select {
+	case cidr = <-r.cidrs:
+	case <-time.After(allocateNextTimeout):
+		err = errCIDRRangeNoCIDRsRemaining
+	}
+	return
+}
+
+func (r *rangeAllocator) Release(cidr *net.IPNet) {
+	clusterIP := r.clusterCIDR.IP.To4()
+	clusterMask := r.clusterCIDR.Mask
+	clusterMaskSize, _ := clusterMask.Size()
+
+	nodeNum := uint32(1) << uint32(r.subNetMaskSize-clusterMaskSize)
+
+	used := (binary.BigEndian.Uint32(clusterIP) ^ binary.BigEndian.Uint32(cidr.IP.To4())) >> uint32(32-r.subNetMaskSize)
+
+	if used > nodeNum {
+		return
+	}
+
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	r.used.SetBit(&r.used, int(used), 0)
+}
+
+// fillCIDRs loops, always searching for the next free CIDR and, if found, fills the CIDRs buffer with it.
+// this func blocks until either there are no remaining free CIDRs, or else the stopCh chan is closed.
+func (r *rangeAllocator) fillCIDRs(stopCh <-chan struct{}) {
+	for {
+		cidr := r.nextFreeCIDR()
+		if cidr == nil {
+			return
+		}
+		select {
+		case <-stopCh:
+			return
+		case r.cidrs <- cidr:
+		}
+	}
+}
+
+// nextFreeCIDR finds a free CIDR, first picking a random cidr, if it is already in use
+// then the CIDR range is scanned sequentially until either a CIDR is found or the scan completes
+// unsuccessfully. an unsuccessful scan returns a nil.
+func (r *rangeAllocator) nextFreeCIDR() *net.IPNet {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	clusterIP := r.clusterCIDR.IP.To4()
+	clusterMask := r.clusterCIDR.Mask
+	clusterMaskSize, _ := clusterMask.Size()
+
+	nodeNum := 1 << uint32(r.subNetMaskSize-clusterMaskSize)
+
+	nextUnused, err := func() (int, error) {
+		j := r.rand.Intn(nodeNum)
```

Comment:
Can you change it to sequential? Allocation problems are much harder to catch if it is random. Take a look at #20048 for an example of where we missed a problem in review.